include/rmm/mr/device/polymorphic_allocator.hpp

```diff
@@ -0,0 +1,264 @@
+  value_type* allocate(std::size_t n, cuda_stream_view stream)
+  {
+    return static_cast<value_type*>(resource()->allocate(n * sizeof(T), stream));
+  }
+
+  /**
+   * @brief Deallocates storage pointed to by `p`.
+   *
+   * `p` must have been allocated from a `rmm::mr::device_memory_resource` `r` that compares equal
+   * to `*resource()` using `r.allocate(n * sizeof(T))`.
+   *
+   * @param p Pointer to memory to deallocate
+   * @param n Number of objects originally allocated
+   * @param stream Stream on which to perform the deallocation
+   */
+  void deallocate(value_type* p, std::size_t n, cuda_stream_view stream)
+  {
+    resource()->deallocate(p, n * sizeof(T), stream);
+  }
+
+  /**
+   * @brief Returns pointer to the underlying `rmm::mr::device_memory_resource`.
+   *
+   * @return Pointer to the underlying resource.
+   */
+  device_memory_resource* resource() const noexcept { return mr_; }
+
+ private:
+  device_memory_resource* mr_{
+    get_current_device_resource()};  ///< Underlying resource used for (de)allocation
+};
+
+template <typename T, typename U>
+bool operator==(polymorphic_allocator<T> const& lhs, polymorphic_allocator<U> const& rhs)
+{
+  return lhs.resource()->is_equal(*rhs.resource());
+}
+
+template <typename T, typename U>
+bool operator!=(polymorphic_allocator<T> const& lhs, polymorphic_allocator<U> const& rhs)
+{
+  return not(lhs == rhs);
+}
+
+/**
+ * @brief Adapts a stream ordered allocator to provide a standard `Allocator` interface
+ *
+ * A stream-ordered allocator (i.e., `allocate/deallocate` use a `cuda_stream_view`) cannot be used
+ *in an interface that expects a standard C++ `Allocator` interface. `stream_allocator_adaptor`
+ *wraps a stream-ordered allocator and a stream to provide a standard `Allocator` interface. The
+ *adaptor uses the wrapped stream in calls the underlying allocator's `allocate` and `deallocate`
+ * functions.
+ *
+ * Example:
+ *\code{c++}
+ * my_stream_ordered_allocator<int> a{...};
+ * cuda_stream_view s = // create stream;
+ *
+ * auto adapted = make_stream_allocator_adaptor(a, s);
+ *
+ * // Allocates storage for `n` int's on stream `s`
+ * int * p = std::allocator_traits<decltype(adapted)>::allocate(adapted, n);
+ *\endcode
+ *
+ * @tparam Allocator Stream ordered allocator type to adapt
+ */
+template <typename Allocator>
+class stream_allocator_adaptor {
+ public:
+  using value_type = typename std::allocator_traits<Allocator>::value_type;
+
+  stream_allocator_adaptor() = delete;
+
+  /**
+   * @brief Construct a `stream_allocator_adaptor` using `a` as the underlying allocator.
+   *
+   * @note: The `stream` must not be destroyed before the `stream_allocator_adaptor`, otherwise
+   * behavior is undefined.
+   *
+   * @param a The stream ordered allocator to use as the underlying allocator
+   * @param stream The stream used with the underlying allocator
+   */
+  stream_allocator_adaptor(Allocator const& a, cuda_stream_view stream) : alloc_{a}, stream_{stream}
```

Comment:
And what happens when the underlying stream gets deleted without the allocator adaptor knowing?

Reply:
```
   * @note: The `stream` must not be destroyed before the `stream_allocator_adaptor`, otherwise
   * behavior is undefined.
```

Reply:
Sorry, missed that! Maybe you need a `set_stream()` so that the adaptor can be kept around if the stream is deleted, just used with a different stream?

Reply:
The primary use case for this adaptor is to pass it to interfaces that don't know anything about streams. I could add a `set_stream()`, but I don't think it would/could ever be used.